### ISSUE TITLE
Generic transaction and block Environment.

### ISSUE THREAD

**Issue by rakita** [bluealloy/revm#916]
After Evm builder is finished we should generalize the Block and Transaction to the trait, this would allow adding new fields to both of those structs and elevate usage of the External Contexts as it would allow multiple generics and easier integration with other types.

Evn should stay as the default type that we use, and this would be overridable in EvmBuilder.

**Comment by rakita** [comment]
Make Transaction generic contains tx type so we have a clean difference between transactions, currently we have done it ad hoc if field is `Some`.

**Comment by Rjected** [comment]
Yeah, at least for the concrete types we would want something like this:
```rust
/// Ethereum tx env
pub struct EthTxEnv {
    // ... the tx env fields for mainnet ethereum
}

/// Data used to calculate the l1 cost of a transaction
pub struct RollupCostData {
    /// The number of ones in the rlp encoding of the signed transaction being executed
    pub ones: u64,
    /// The number of zeroes in the rlp encoding of the signed transaction being executed
    pub zeroes: u64,
}

/// Optimism tx env
pub struct OptimismTxEnv {
    /// The underlying ethereum tx environment
    pub eth_env: EthTxEnv,
    // ... then the optimism specific fields
    // ideally including this:
    pub rollup_cost_data: RollupCostData,
}
```
And each of these would implement an environment trait with methods to expose what we currently use in mainnet eth execution.

We'd also want some flexibility, for example to let optimism actually use its custom tx env fields.
For example here are some things that optimism currently does with custom fields:
https://github.com/bluealloy/revm/blob/7cf937ae2e902c0eb6b2d2e6c385903a3d65851e/crates/primitives/src/env.rs#L84-L94
https://github.com/bluealloy/revm/blob/7cf937ae2e902c0eb6b2d2e6c385903a3d65851e/crates/primitives/src/env.rs#L207-L212
https://github.com/bluealloy/revm/blob/7cf937ae2e902c0eb6b2d2e6c385903a3d65851e/crates/revm/src/optimism/handler_register.rs#L125-L150

Not sure how yet we could provide this flexibility, maybe adding some hooks that are run in the beginning of tx validation, or elsewhere

**Comment by rakita** [comment]
This is imagined to be a `TxTrait` or `TxEnvTrait` that you can override if needed. It is up to lib user to specify different generics of the trait and add handlers that use that trait. In reth, this would be just implementing `TxTrait` over &Transaction and you could use `&Transaction` directly inside `revm` without need to populate `Env`

**Comment by Autoparallel** [comment]
I'll give this a shot

**Comment by rakita** [comment]
Done on main
